### DISCUSSION TITLE
Moving/swapping entire workspace?

### DISCUSSION THREAD

**Discussion by ttencate** [i3/i3#4969] [Q&A] [answered]
Context: I'm working on several projects and often alternating between them. I like to keep the windows for a particular project open when switching to another one.

I have Alt+F1...F10 set up to switch to workspace 1...10, but only F1...F3 are really in an ergonomic position. And also, these are already in muscle memory. So I'd like to have my current project on those workspaces, and quickly throw any background projects onto F4...F10. I have Win+F1...F10 bound to move windows to another workspace.

Right now, I switch projects by a sequence of e.g. Alt+F1, Win+F4, Alt+F1, Win+F4, ... until the workspace is empty and ready for another project. But this is pretty tedious, and also it doesn't preserve workspace layout.

What I'd really like is a way to swap the current workspace, including layout, with another. Is this possible?

**Comment by boyvlad85** [comment]
https://i3wm.org/docs/userguide.html#workspace_auto_back_and_forth

**Comment by ttencate** [comment]
@boyvlad85 Thanks, I didn't know about that option, sounds useful! It doesn't help address my current issue though. I'm trying to move the entire contents of a workspace to another workspace, not just trying to switch between workspaces.

**Comment by budRich** [comment, marked as answer]
You can probably achieve this by using some `workspace rename` shenanigans.  
Figure out what the next free and active workspace number is by parsing output of `i3-msg -t get_workspaces` , and then:  
`i3-msg rename workspace to "$next_free_workspace"; workspace "$active_workspace"`
I guess this only changes the actual workspace **name** and not the number, maybe?

There was no fancy link for the topic in in the userguide, but it is there, and below:

6.8.2. Renaming workspaces
You can rename workspaces. This might be useful to start with the default numbered workspaces, do your work, and rename the workspaces afterwards to reflect what’s actually on them. You can also omit the old name to rename the currently focused workspace. This is handy if you want to use the rename command with i3-input.

Syntax:

rename workspace <old_name> to <new_name>
rename workspace to <new_name>
Examples:

i3-msg 'rename workspace 5 to 6'
i3-msg 'rename workspace 1 to "1: www"'
i3-msg 'rename workspace "1: www" to "10: www"'
i3-msg 'rename workspace to "2: mail"'
bindsym $mod+r exec i3-input -F 'rename workspace to "%s"' -P 'New name: '
If you want to rename workspaces on demand while keeping the navigation stable, you can use a setup like this:

Example:

bindsym $mod+1 workspace number "1: www"
bindsym $mod+2 workspace number "2: mail"
...
If a workspace does not exist, the command workspace number "1: mail" will create workspace "1: mail".

If a workspace with number 1 does already exist, the command will switch to this workspace and ignore the text part. So even when the workspace has been renamed to "1: web", the above command will still switch to it.

**Reply by orestisfl** [reply]
The link: https://i3wm.org/docs/userguide.html#_renaming_workspaces

**Comment by ttencate** [comment]
Wow @budRich, that's great, thank you! `rename workspace to` does exactly what I need, with the following config (with `set $win Mod4` at the top):

```
# move current workspace to another number
bindsym Ctrl+$win+F1 rename workspace to 1
bindsym Ctrl+$win+F2 rename workspace to 2
bindsym Ctrl+$win+F3 rename workspace to 3
bindsym Ctrl+$win+F4 rename workspace to 4
bindsym Ctrl+$win+F5 rename workspace to 5
bindsym Ctrl+$win+F6 rename workspace to 6
bindsym Ctrl+$win+F7 rename workspace to 7
bindsym Ctrl+$win+F8 rename workspace to 8
bindsym Ctrl+$win+F9 rename workspace to 9
bindsym Ctrl+$win+F10 rename workspace to 10
```

**Comment by alanbarnett** [comment]
For something more granular, I have `Mod+a` set to `focus parent` and `Mod+Shift+a` set to `focus child`. This lets you "zoom out" your focus in a sense, so that you focus on groups of windows and eventually the entire workspace. For example, if I have some tabbed windows next to two windows in a vertical split, and I want them to trade sides, if you `focus parent` on any individual window it will go up a level and focus on the group of them. So, the tabbed windows would all be highlighted, or if you were focused on the split, the two split windows would be highlighted. Then you move them like normal to change sides. Using this, it's very easy to focus on all of the windows in a workspace, and then move them to a different workspace in a group.

Edit: Also, this is also very convenient for opening splits in odd places without needing to rearrange and resize everything again. If you highlight everything and open a split below, that's easier than making a window, moving it down, and then trying to shove everything else back on top of it.

**Reply by ttencate** [reply]
Those are both excellent tips! I'll train myself to use them.

**Comment by mutageneral** [comment]
Rename current workspace to any name using an interactive dmenu menu with a 2 line shell script:
```sh
#!/bin/sh
name=$(echo "" | dmenu -p "rename workspace to:")
i3-msg -- rename workspace to \""$name"\"
```
You can then call this script from any keybinding you like. I prefer this over keybinding to rename the current workspace to a specific number as I do not rename workspace often, so they don't eat up keybinding that can be used for more useful actions.

Also, you can rename it to whatever you want. This means if I'm already using all of my 13 workspaces, I can basically create new ones on the fly. For example, switch to and create the workspace "13", open a program, rename the workspace to "other" or whatever you want, switch to and create the workspace "13" again. You now have infinite workspaces.

You can also give "tags" or "nicknames" to existing workspaces while keeping their original number and number keybinding function. If you rename workspace "2" to "2 : web", it will still function with the keybindings involving the "2" key. If you renamed it to "other", it would not, and "2" would be created again next time you use a keybindings involving the "2" key.
